lib/Conversion/TorchToLinalg/DataMovement.cpp

```diff
@@ -1834,6 +1834,147 @@ class ConvertAtenViewAsRealOp : public OpConversionPattern<AtenViewAsRealOp> {
 };
 } // namespace
 
+namespace {
+class ConvertAtenDiagonalOp : public OpConversionPattern<AtenDiagonalOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(AtenDiagonalOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+
+    int64_t offset;
+    if (!matchPattern(op.getOffset(), m_TorchConstantInt(&offset)))
+      return rewriter.notifyMatchFailure(op, "offset must be constant");
+    int64_t dim1;
+    if (!matchPattern(op.getDim1(), m_TorchConstantInt(&dim1)))
+      return rewriter.notifyMatchFailure(op, "dim1 must be constant");
+    int64_t dim2;
+    if (!matchPattern(op.getDim2(), m_TorchConstantInt(&dim2)))
+      return rewriter.notifyMatchFailure(op, "dim2 must be constant");
+
+    Value inputMatrix = adaptor.getSelf();
+    RankedTensorType inType = inputMatrix.getType().cast<RankedTensorType>();
+    int64_t inputRank = inType.getRank();
+    RankedTensorType outType = getTypeConverter()
+                       ->convertType(op->getResult(0).getType())
+                       .cast<RankedTensorType>();
+    Type elementType = inType.getElementType();
+    Location loc = op.getLoc();
+
+    if (inputRank < 2) 
+      return rewriter.notifyMatchFailure(op, "input must have at least two dimensions");
+
+    dim1 = toPositiveDim(dim1, inputRank);
+    if (!isValidDim(dim1, inputRank))
+      return rewriter.notifyMatchFailure(op, "dim1 out of range");
+    dim2 = toPositiveDim(dim2, inputRank);
+    if (!isValidDim(dim2, inputRank))
+      return rewriter.notifyMatchFailure(op, "dim2 out of range");
+    if (dim1 == dim2) 
+      return rewriter.notifyMatchFailure(op, "diagonal dimensions cannot be identical");
+
+    bool offsetNegative = offset < 0;
+    Value dim1Size, dim2Size;
+    dim1Size = getDimOp(rewriter, loc, adaptor.getSelf(), dim1);
```

Comment:
There is a  `getTensorSizes` function to get all sizes at once. `inputSizes[dim]` is  a bit more readable than using getDimOp in a complex expression as you do below.

Reply:
Thanks for the hint, I could change this together with other modifications.